lib/spack/spack/spec_parser.py

```diff
@@ -360,25 +361,34 @@ def _parse_node(self, root_spec):
             raise spack.spec.RedundantSpecError(root_spec, "^" + str(dependency))
         return dependency, parser_warnings
 
-    def _apply_toolchain(self, spec: "spack.spec.Spec", toolchain_config: Union[str, List]):
+    def _apply_toolchain(self, spec: "spack.spec.Spec", name: str):
+        toolchain = self.parsed_toolchains.get(name, None)
+        if toolchain:
+            spec.constrain(toolchain)
+            return
+
+        toolchain_config = self.toolchains[name]
+
         # Single string entries constrain the spec
         if isinstance(toolchain_config, str):
             toolchain = parse_one_or_raise(toolchain_config)
             self._ensure_all_direct_edges(toolchain)
+            self.parsed_toolchains[name] = toolchain
             spec.constrain(toolchain)
             return
 
         # List entries we apply each list element
         for entry in toolchain_config:
-            constraint = parse_one_or_raise(entry["spec"])
+            toolchain = parse_one_or_raise(entry["spec"])
             when = entry.get("when", "")
-            self._ensure_all_direct_edges(constraint)
+            self._ensure_all_direct_edges(toolchain)
 
             # Conditions are applied to every edge in the constraint
-            for edge in constraint.traverse_edges():
+            for edge in toolchain.traverse_edges():
                 edge.when.constrain(when)
 
-            spec.constrain(constraint)
+            self.parsed_toolchains[name] = toolchain
```

Comment:
I think this is wrong _within_ the loop, the cached toolchain seems to be now only the last 
```yaml
- spec: ...
  when: ...
```
entry. I'll fix that in a following commit.

Reply:
Oh ugh yeah that's definitely wrong. Let me know if you want me to fix it instead.